base/compiler/optimize.jl

```diff
@@ -34,7 +34,7 @@ struct InliningState{S <: Union{EdgeTracker, Nothing}, T <: Union{InferenceCache
 end
 
 mutable struct OptimizationState
-    linfo::MethodInstance
+    linfo::Union{MethodInstance, Nothing}
```

Comment:
Would it add clarity to define a new singleton to use a signal something is a OpaqueClosure, 
rather than using `Nothing` for this?
Or is using `nothing` a sitatution that in theory could occur more widely than just for OpaqueClosures?

Or from another direction, should there be a `OpaqueClosureOptimizationState` that doens't have this field?

Reply:
Yeah, that might be cleaner, but honestly this whole thing needs a refactor, but that's for another PR.